include/zephyr/pm/pm.h

```diff
@@ -53,18 +53,119 @@ extern "C" {
+	 * Optional callback context
 	 */
-	void (*state_exit)(enum pm_state state);
+	const void *ctx;
 };
 
+/**
+ * @brief Helper macro for building a pm_notifiers list name.
+ *
+ * @param node_id zephyr,power-state node identifer
+ */
+#define Z_PM_NOTIFIERS_DT_NAME(node_id) UTIL_CAT(pm_notifiers_, DT_NODE_CHILD_IDX(node_id))
+
+/**
+ * @brief Generate a pm_notifiers list for the passed power-states child node
+ *	  identifier.
+ *
+ * @param node_id zephyr,power-state node identifier
+ */
+#define Z_PM_NOTIFIERS_CREATE_DT_STATES(node_id)			\
+	static sys_slist_t Z_PM_NOTIFIERS_DT_NAME(node_id) =		\
+		SYS_SLIST_STATIC_INIT(&Z_PM_NOTIFIERS_DT_NAME(node_id))
+
+/**
+ * @brief Provides a pm_notifiers list pointer for a power-states child node
+ *	  identifier.
+ *
+ * @param node_id zephyr,power-state node identifier
+ */
+#define Z_PM_NOTIFIERS_DT_STATES(node_id) &Z_PM_NOTIFIERS_DT_NAME(node_id)
+
+#ifdef CONFIG_PM
+/**
+ * @brief Generic helper for defining a pm_notifier struct.
+ *
+ * @param id Unique alphanumeric identifier, specific to usage context
+ *	     (avoid using only numbers here to prevent collisions)
+ * @param state_direction Transition direction: state entry, exit, or both
+ * @param callback_fn callback function
+ */
+#define PM_NOTIFIER_DEFINE(id, state_direction, callback_fn, context)	\
+	static struct pm_notifier UTIL_CAT(pm_notifier_, id) = {	\
+		.direction = state_direction,				\
+		.callback = callback_fn,				\
+		.ctx = context,						\
+	}
+
+/**
+ * @brief Retrieve the name of a pm_notifier struct instance.
+ *
+ * @param id Unique alphanumeric identifier, specific to usage context
+ */
+#define PM_NOTIFIER(id) UTIL_CAT(pm_notifier_, id)
+#else
+#define PM_NOTIFIER_DEFINE(id, state_direction, callback_fn, context)
+#define PM_NOTIFIER(id) NULL
+#endif
+
+/**
+ * @brief Driver helper for defining a per-node pm_notifier struct.
+ *
+ * @param node_id Node identifier
+ * @param state_direction Transition direction: state entry, exit, or both
+ * @param callback_fn Callback function
+ */
+#define PM_NOTIFIER_DT_DEFINE(node_id, state_direction, callback_fn)	\
+	PM_NOTIFIER_DEFINE(DT_DEP_ORD(node_id), state_direction,	\
+			   callback_fn, DEVICE_DT_GET(node_id))
+
+/**
+ * @brief Driver helper for defining a per-instance pm_notifier struct.
+ *
+ * @param inst Instance number
+ * @param state_direction Transition direction: state entry, exit, or both
+ * @param callback_fn Callback function
+ */
+#define PM_NOTIFIER_DT_INST_DEFINE(inst, state_direction, callback_fn)	\
+	PM_NOTIFIER_DEFINE(DT_INST_DEP_ORD(inst), state_direction,	\
+			   callback_fn, DEVICE_DT_INST_GET(inst))
+
+/**
+ * @brief Retrieve a node's pm_notifier struct.
+ *
+ * @param node_id Node identifier
+ */
+#define PM_NOTIFIER_DT(node_id) PM_NOTIFIER(DT_DEP_ORD(node_id))
+
+/**
+ * @brief Retrieve an instance's pm_notifier struct.
+ *
+ * @param inst Instance number
+ */
+#define PM_NOTIFIER_DT_INST(inst) PM_NOTIFIER(DT_INST_DEP_ORD(inst))
+
```

Comment:
I particularly don't see a reason to have this specialization here, if one needs using it in devices, a notifier with the device as context can easily be created with the device. Also, these macros are confusing IMHO since they don't tell anything about having a device as context.

Reply:
I suppose that applies to all the `_DT` macros added here. IMHO they help driver/app code look cleaner, though naming could be improved.

Would something like `PM_NOTIFIER_DT_DEVICE_INST` provide enough information that the device is passed/present in context?

Reply:
Adapted macro naming, please have a look.

Reply:
What I was saying is to have only `PM_NOTIFIER_DEFINE` and let what is done in these macros to the driver to do. Further if that become re-current we introduce these helpers, I would avoid adding more APIs than needed for now.